### PR TITLE
WT-6798 Utilize Arm LSE atomics and the correct strength barriers

### PR DESCRIPTION
**Pull request by AGSaidi**
Armv8 originally had load-exclusive/store-exclusive atomics. In Arm v8.1 these were extended with the Large Scale Extensions (LSE) which provided native atomic primitives such as compare-and-swap, load-and-add, etc which have been shown to scale much better on larger core-count systems. These are the default for -march=armv8.1-a, but to retain compatibility with older Arm systems the -moutline-atomics option is available which includes both new and old-style atomics and chooses dynamically. moutline-atomics is the default in gcc10 and has been backported to some distro gcc7 compilers as well as the official gcc 8 and 9 branches. Detect if the flag is available and if it is use it. 

bench/wtperf/runners/medium-btree.wtperf ops/s increases by 5x with  the flag applied.

Additionally, the barriers used for Arm are stronger than required. The existing dsb barriers guarantee completion, not ordering and synchronize other system activities such as tlb and cache maintenance instructions. Additionally, the sharability domain of them is system. Both of these are stronger than required. A dmb enforces ordering described in the header and the stronger dsb isn't needed. Furthermore, given that all the entities in the system which are depending on the ordering are CPUs the shareability doesn't need to order accesses with respect to device or non-cacheable memory, so it can be reduced to the inner-sharable domain.

This improve performance on the above test by about 4%. 

Both changes passed make check.

### CODE REVIEW

**Comment by lukech** [comment]
autobuild: Can one of the admins verify this patch?

**Comment by tetsuo-cpp** [comment]
Hey @AGSaidi, nice to meet you. Thanks for submitting this patch!

Before we go ahead and get this merged, could you please sign our [Contributor Agreement](https://www.mongodb.com/legal/contributor-agreement)?

**Comment by AGSaidi** [comment]
CLA signed

**Comment by AGSaidi** [comment]
@tetsuo-cpp anything else you need now that i've signed the SLA?

**Comment by tetsuo-cpp** [comment]
Sorry for the wait @AGSaidi.
I'm waiting to hear back internally about the CLA signing confirmation. I don't expect any issues.

In the interim, I'll look into getting a hold of an ARM machine to do some stress testing with this changeset and do some reading about LSE to improve my understanding. I might ping you with some questions if that's ok.

Thanks for your patience.

**Comment by AGSaidi** [comment]
Thanks @tetsuo-cpp please let me know if I can answer any questions or help with getting ahold of arm64 hw.

**Comment by tetsuo-cpp** [comment]
evergreen retry

**Comment by tetsuo-cpp** [comment]
@AGSaidi 
I got the CLA confirmation overnight. 👍 

@agorrod 
Do you mind taking a look at this?

I ran some tests on an Evergreen ARM machine and things are looking good.

- I can confirm that I see a nice perf improvement for the few `wtperf` jobs that I tried. I think some GCC versions (I was running on 9) use this option by default so I had to explicitly turn it off with `-mno-outline-atomics` to see a perf difference.
- Python tests and `make check` pass. I also ran a `format` stress test which completed ~140 runs without incident.
- I'm going to kick off a MongoDB patch build specifically for the ARM variants.

**Review by tetsuo-cpp — APPROVED** [review]
LGTM.

**Comment by AGSaidi** [comment]
> Some versions have it default

Yes, in GCC10 it's the default. There are backports of the GCC10 feature to 7, 8, 9 where it's not the default. 

> I'm going to kick off a MongoDB patch build specifically for the ARM variants.

Apologies, i haven't looked into it and it was going to be the next thing after this got merged, but is wired tiger built separately and linked into MongoDB or would a similar patch be required for the MongoDB repo?

**Comment by tetsuo-cpp** [comment]
> Apologies, i haven't looked into it and it was going to be the next thing after this got merged, but is wired tiger built separately and linked into MongoDB or would a similar patch be required for the MongoDB repo?

We periodically synchronize what's in our `develop` branch with MongoDB's source tree. So once this PR gets merged, this changeset will end up in MongoDB `master` within a day or two and will end up in future MongoDB releases.

File: src/include/gcc.h
Comment (agorrod) [inline src/include/gcc.h:206]: It would be great if you can add the content in the description on the PR as a comment above here, so we can remember why the `ish` operators are enough.

**Comment by tetsuo-cpp** [comment]
@AGSaidi 
I noticed that this patch is triggering some lint rules. Since your fork is on `master`, you'll have to rebase onto `develop` and then run `dist/s_all` locally in your tree (`develop` has updated linting rules so the difference is important).

Alternatively, you can apply this patch that I made which should apply the correct changes:
```
commit ef5812a8f48742e35649a246807283c28980f47e
Author: Alex Cameron <alex.cameron@10gen.com>
Date:   Thu Oct 15 23:52:58 2020 +0000

    Lint fixes.

diff --git a/dist/s_string.ok b/dist/s_string.ok
index be6ae0893..3df400be1 100644
--- a/dist/s_string.ok
+++ b/dist/s_string.ok
@@ -698,12 +698,14 @@ dll
 dlopen
 dlsym
 dmalloc
+dmb
 dmsg
 doxgen
 doxygen
 drealloc
 ds
 dsb
+dsbs
 dsk
 dsrc
 dst
@@ -894,6 +896,9 @@ isalpha
 iscntrl
 isdigit
 isgraph
+ish
+ishld
+ishst
 islocked
 islower
 ispo
@@ -1224,6 +1229,7 @@ setstr
 setv
 setvbuf
 sfence
+shareability
 signalled
 sii
 sizeof
@@ -1306,6 +1312,7 @@ timestamp
 timestamped
 timestamps
 tinfo
+tlb
 tmp
 todo
 tokenizer
diff --git a/src/include/gcc.h b/src/include/gcc.h
index 2568972b3..85ba3214d 100644
--- a/src/include/gcc.h
+++ b/src/include/gcc.h
@@ -220,26 +220,25 @@ WT_ATOMIC_FUNC(size, size_t, size_t *vp, size_t v)
 #define WT_PAUSE() __asm__ volatile("yield" ::: "memory")

 /*
- * dmb are chosen here because they are sufficient to guarantee the ordering
- * described above. We don't want to use dsbs because they provide a much
- * stronger guarantee of completion which isn't required. Additionally, dsbs
- * synchronize other system activities such as tlb and cache maintenance
- * instructions which is not required in this case.
+ * dmb are chosen here because they are sufficient to guarantee the ordering described above. We
+ * don't want to use dsbs because they provide a much stronger guarantee of completion which isn't
+ * required. Additionally, dsbs synchronize other system activities such as tlb and cache
+ * maintenance instructions which is not required in this case.
  *
- * A sharability domain of inner-sharable is selected beacause all the
- * entities participating in the ordering requirements are CPUs and ordering
- * with respect to other devices or memory-types isn't required.
+ * A shareability domain of inner-shareable is selected because all the entities participating in
+ * the ordering requirements are CPUs and ordering with respect to other devices or memory-types
+ * isn't required.
  */
-#define WT_FULL_BARRIER()                        \
-    do {                                         \
+#define WT_FULL_BARRIER()                         \
+    do {                                          \
         __asm__ volatile("dmb ish" ::: "memory"); \
     } while (0)
-#define WT_READ_BARRIER()                        \
-    do {                                         \
+#define WT_READ_BARRIER()                           \
+    do {                                            \
         __asm__ volatile("dsb ishld" ::: "memory"); \
     } while (0)
-#define WT_WRITE_BARRIER()                       \
-    do {                                         \
+#define WT_WRITE_BARRIER()                          \
+    do {                                            \
         __asm__ volatile("dsb ishst" ::: "memory"); \
     } while (0)

```

**Comment by AGSaidi** [comment]
@tetsuo-cpp  rebased onto develop and ran `cd dist/ && ./s_all` as requested. Squashed the resulting changes together.

**Comment by AGSaidi** [comment]
> We periodically synchronize what's in our develop branch with MongoDB's source tree. So once this PR gets merged, this changeset will end up in MongoDB master within a day or two and will end up in future MongoDB releases

Does MongoDB use the wiredtiger make file to build, and thus the `-moutline-atomics` flag will be used, or do I need to create a change to MongoDB as well? I don't know if any of the `Boost::atomic` code is used, but it might be worth detecting and applying the flag there too.

**Comment by tetsuo-cpp** [comment]
> @tetsuo-cpp rebased onto develop and ran `cd dist/ && ./s_all` as requested. Squashed the resulting changes together.

Looking good. Thanks for that.

> Does MongoDB use the wiredtiger make file to build, and thus the `-moutline-atomics` flag will be used, or do I need to create a change to MongoDB as well? I don't know if any of the `Boost::atomic` code is used, but it might be worth detecting and applying the flag there too.

Ah, sorry now I understand what you mean. Yeah, WiredTiger is built and linked separately in the MongoDB build. MongoDB has its own `SConstruct` file that it uses to manage compiler flags and things like that.

I'm not super familiar with their build setup but I can file a ticket to get someone to look into it if you'd like.

**Comment by tetsuo-cpp** [comment]
There are a few typos in the comment that are still getting picked up by the linter.

```
beacause => because
sharable => shareable
sharability => shareability
```

Sorry, I'd make the fixes directly but I don't have write access to your fork.

**Review by agorrod — APPROVED** [review]
Thanks so much for your contributions - LGTM

**Comment by tetsuo-cpp** [comment]
I've just merged the patch. Thanks for your patience @AGSaidi.

I've filed a ticket to check whether we're using this compiler flag in MongoDB [here](https://jira.mongodb.org/browse/SERVER-51722). I'll keep an eye on this ticket but if you're interested in the progress, feel free to "watch" it to get updates.

**Comment by AGSaidi** [comment]
thanks @tetsuo-cpp!

**Comment by AGSaidi** [comment]
@tetsuo-cpp A comment on `-moutline-atomics`. The compiler does the right thing here, so if you specify a version of the architecture that includes the large-scale-extensions (LSE) such as armv8.1-a then -moutline-atomics becomes a nop and only the LSE instructions are emitted.

**Pull request merged**
